experimental/oak_baremetal_channel/SPEC.MD

```diff
@@ -0,0 +1,124 @@
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119).
+
+## Method
+
+Methods are exposed by the server. They MAY be remotely invoked by the Unrusted
+Loader. Each method has clearly defined parameters and return types. The Oak IDL
+is used to define which methods are available, their parameter and return types,
+as well the protcol used to encode parameters and return values to bytes.
+
+## Invocation
+
+An invocations of a Method. Each invocation consists out of a Request Message
+followed by a Response Message.
+
+### Issuing an invocation
+
+The client MAY issue an invocation by sending a Request Message.
+
+The server MUST NOT send Request Message.
+
+### Handling an invocation
+
+The server MUST respond to each Request Message with a single Response Message.
+It SHOULD respond to Request Messages in first-in-first-out order, but MAY
+prioritize some requests. The server MAY issue Response Messages before the
+Request Message has been received in its entirety. Sending a Response Messages
+concludes the invocation.
+
+The client MUST NOT send Response Messages.
+
+## Messages
+
+A message forms one half of an Invocation.
+
+### Request Message Definition
+
+A byte encoded request message MUST consist of the following seqeuential fields:
+
+- length, u32, little endian: length of the entire message, including this field
+- invocation_id, u32, little endian: Identifies the invocation the request
+  initiates
+- method_id, u32, little endian: identifies the method that is being invoked
+- padding, 4 bytes: reserved to maintain 8 byte alignment. MUST be unset and
+  ignored
+- body, variable length byte array: byte encoded method parameters
+
+### Response Message Definition
+
+A byte encoded response message MUST consist of the following seqeuential
+fields:
+
+- length, u32, little endian: length of the entire message, including this field
+- invocation_id, u32, little endian: MUST match the invocation_id of the
+  relevant Request Message
+- status_code, u32, little endian: response status as a
+  [gRPC status code](https://grpc.github.io/grpc/core/md_doc_statuscodes.html)
+- status_message_length, u32, little endian: length of the status message
+- status_message, utf-8 encoded string: developer-facing error message
+- body, variable length byte array: byte encoded method return value
+
+### Sending Messages
+
+To send a message, it MUST first be encoded into frames. The resulting frame set
+MUST be sent as an uninterrupted sequence.
+
+### Receiving Messages
+
+To receive a message the recipent incrementaly reads frames from the
+communication channel. The recipent MAY start parsing a message before all of
+its frames have been received.
+
+### Encoding Messages into Frames
+
+The byte representation of a message MUST be chunked. The last chunk MAY be
```

Comment:
Can we avoid introducing the term chunk here, or define it?

Reply:
Up to you, maybe it's fine as it is, if we assume it means "the frame body"?

Reply:
Replaced chunks with frame bodies when used a noun, agreed that it's clearer. :-)